windows/ReactNativeAsyncStorage/DBStorage.cpp

```diff
@@ -0,0 +1,451 @@
+    }
+
+    // Checks if sqliteResult is SQLITE_OK. If not, reports the error via
+    // callback & returns false.
+    bool CheckSQLiteResult(sqlite3* db, const DBStorage::Callback& callback, int sqliteResult) {
+        if (sqliteResult == SQLITE_OK) {
+            return true;
+        }
+        else {
+            InvokeError(callback, sqlite3_errmsg(db));
+            return false;
+        }
+    }
+
+    using Statement = std::unique_ptr<sqlite3_stmt, decltype(&sqlite3_finalize)>;
+
+    // Creates a prepared SQLite statement. On error, returns nullptr
+    Statement PrepareStatement(sqlite3* db, DBStorage::Callback& callback, const char* stmt) {
+        sqlite3_stmt* pStmt{ nullptr };
+        if (!CheckSQLiteResult(db, callback, sqlite3_prepare_v2(db, stmt, -1, &pStmt, nullptr))) {
+            return { nullptr, sqlite3_finalize };
+        }
+        return { pStmt, &sqlite3_finalize };
+    }
+
+    // Binds the index-th variable in this prepared statement to str.
+    bool BindString(
+        sqlite3* db,
+        const DBStorage::Callback& callback,
+        const Statement& stmt,
+        int index,
+        const std::string& str) {
+        return CheckSQLiteResult(db, callback, sqlite3_bind_text(stmt.get(), index, str.c_str(), -1, SQLITE_TRANSIENT));
+    }
+
+} // namespace
+
+DBStorage::DBStorage() {
+    auto const localAppDataParh = winrt::ApplicationData::Current().LocalFolder().Path();
+    std::wstring wPath(localAppDataParh.data());
+    wPath += L"\\AsyncStorage.db";
+    auto const path = ConvertWstrToStr(wPath);
+    if (sqlite3_open_v2(
+        path.c_str(),
+        &m_db,
+        SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_FULLMUTEX,
+        nullptr) != SQLITE_OK) {
+        auto exception = std::runtime_error(sqlite3_errmsg(m_db));
+        sqlite3_close(m_db);
+        throw exception;
+    }
+
+    int userVersion = 0;
+    auto getUserVersionCallback = [](void* pv, int cCol, char** rgszColText, char** /*rgszColName*/) {
+        if (cCol < 1) {
+            return 1;
+        }
+        *static_cast<int*>(pv) = atoi(rgszColText[0]);
+        return SQLITE_OK;
+    };
+
+    Exec(m_db, u8"PRAGMA user_version", getUserVersionCallback, &userVersion);
+
+    if (userVersion == 0) {
+        Exec(
+            m_db,
+            u8"CREATE TABLE IF NOT EXISTS AsyncLocalStorage(key TEXT PRIMARY KEY, value TEXT NOT NULL); PRAGMA user_version=1");
+    }
+}
+
+DBStorage::~DBStorage() {
+    decltype(m_action) action;
+    {
+        // If there is an in-progress async task, cancel it and wait on the
+        // condition_variable for the async task to acknowledge cancellation by
+        // nulling out m_action. Once m_action is null, it is safe to proceed
+        // wth closing the DB connection
+        winrt::slim_lock_guard guard{ m_lock };
+        m_tasks.clear();
+        if (m_action) {
+            m_action.Cancel();
+            m_cv.wait(m_lock, [this]() { return m_action == nullptr; });
```

Comment:
Do we know what kind of apartment that `DBStorage`'s destructor will run in? We're blocking here without pumping messages, which could lead to deadlocks in general for a STA or ASTA thread (for a MTA thread, it's just impolite. 🙂)

Reply:
I'm a little bit fuzzy here so correct me if I'm wrong...

We used it as a stack variable in RNCAsyncStorage, so the destructor should be called when the AsyncStorage module gets destructed. DBStorage is not a public type so outside users won't be able to touch it. So I *think* we are good?

Reply:
That sounds right, but it just shifts the questions up a level... what kind of thread is the `AsyncStorage` module being destroyed on?

Reply:
I changed it to shared lock like you did in the win32 module, which I think should avoid the deadlock.

Reply:
Hmm, no, that's not what I did in the win32 module - ping me on Teams tomorrow during working hours and I'll explain my question here a little more fully.